crypto/cmp/cmp_server.c

```diff
@@ -0,0 +1,656 @@
+
+    OSSL_CMP_PKISI_free(si);
+    return msg;
+
+ err:
+    OSSL_CMP_PKISI_free(si);
+    return NULL;
+}
+
+static OSSL_CMP_MSG *process_rr(OSSL_CMP_SRV_CTX *srv_ctx, OSSL_CMP_MSG *req)
+{
+    OSSL_CMP_MSG *msg;
+    OSSL_CMP_REVDETAILS *details;
+    OSSL_CRMF_CERTID *certId;
+    OSSL_CRMF_CERTTEMPLATE *tmpl;
+    X509_NAME *issuer;
+    ASN1_INTEGER *serial;
+
+    if (!ossl_assert(srv_ctx != NULL && req != NULL))
+        return NULL;
+
+    if ((details = sk_OSSL_CMP_REVDETAILS_value(req->body->value.rr,
+                                                OSSL_CMP_REVREQSID)) == NULL) {
+        CMPerr(0, CMP_R_ERROR_PROCESSING_MSG);
+        return NULL;
+    }
+
+    /* accept revocation only for the certificate we send in ir/cr/kur */
+    tmpl = details->certDetails;
+    issuer = OSSL_CRMF_CERTTEMPLATE_get0_issuer(tmpl);
+    serial = OSSL_CRMF_CERTTEMPLATE_get0_serialNumber(tmpl);
+    if (X509_NAME_cmp(issuer, X509_get_issuer_name(srv_ctx->certOut)) != 0
+            || ASN1_INTEGER_cmp(serial,
+                                X509_get0_serialNumber(srv_ctx->certOut)) != 0) {
+        CMPerr(0, CMP_R_REQUEST_NOT_ACCEPTED);
+        return NULL;
+    }
+
+    if ((certId = OSSL_CRMF_CERTID_gen(issuer, serial)) == NULL)
+        return NULL;
+
+    if ((msg = ossl_cmp_rp_new(srv_ctx->ctx, srv_ctx->pkiStatusOut, certId,
+                               srv_ctx->sendUnprotectedErrors)) == NULL)
+        CMPerr(0, CMP_R_ERROR_CREATING_RR);
+    OSSL_CRMF_CERTID_free(certId);
+    return msg;
+}
+
+static OSSL_CMP_MSG *process_certConf(OSSL_CMP_SRV_CTX *srv_ctx,
+                                      OSSL_CMP_MSG *req)
+{
+    OSSL_CMP_CTX *ctx = srv_ctx->ctx;
+    OSSL_CMP_MSG *msg = NULL;
+    OSSL_CMP_CERTSTATUS *status = NULL;
+    ASN1_OCTET_STRING *tmp = NULL;
+    int res = -1;
+    int num = sk_OSSL_CMP_CERTSTATUS_num(req->body->value.certConf);
+
+    if (num == 0) {
+        OSSL_CMP_err(ctx, "certificate rejected by client");
+    } else {
+        if (num > 1)
+            OSSL_CMP_warn(ctx, "All CertStatus but the first will be ignored");
+        status = sk_OSSL_CMP_CERTSTATUS_value(req->body->value.certConf,
+                                              OSSL_CMP_CERTREQID);
+    }
+
+    if (status != NULL) {
+        /* check cert request id */
+        if (ossl_cmp_asn1_get_int(status->certReqId) != srv_ctx->certReqId) {
+            /* in case of error, invalid reqId -1 */
+            CMPerr(0, CMP_R_UNEXPECTED_REQUEST_ID);
+            return NULL;
+        }
+
+        /* check cert hash by recalculating it in place */
+        tmp = status->certHash;
+        status->certHash = NULL;
+        if (ossl_cmp_certstatus_set_certHash(status, srv_ctx->certOut))
+            res = status->certHash == NULL ? 0 /* avoiding SCA false positive */
+                : ASN1_OCTET_STRING_cmp(tmp, status->certHash) == 0;
+        ASN1_OCTET_STRING_free(status->certHash);
+        status->certHash = tmp;
```

Comment:
This seems fairly horrible. We seem to re-using the certHash value in status as a temporary place to store a value, just to resurrect it afterwards again. Is there no way to get the certHash without setting it in the status first?

Reply:
I agree this is ugly. 
We'll improve that, which will likely require refactoring `ossl_cmp_certstatus_set_certHash()`.

Reply:
see #206